pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,818 @@
+// It maintains indexes between groups and entities to make the query efficient. It supports callers to register
+// callbacks that will be called when a specific type of groups' entities are updated.
+type Interface interface {
+	// AddGroup adds or updates a group to the index. The caller can then get entities selected by this group.
+	AddGroup(groupType GroupType, name string, selector *types.GroupSelector)
+	// DeleteGroup deletes a group from the index.
+	DeleteGroup(groupType GroupType, name string)
+	// AddEventHandler registers an eventHandler for the given type of groups. When any Pod/ExternelEntity/Namespace
+	// update affects the given kind of groups, the eventHandler will be called with the affected groups.
+	AddEventHandler(groupType GroupType, handler eventHandler)
+	// GetEntities returns the selected Pods or ExternalEntities for the given group.
+	GetEntities(groupType GroupType, name string) ([]*v1.Pod, []*v1alpha2.ExternalEntity)
+	// GetGroupsForPod returns the groups that select the given Pod.
+	GetGroupsForPod(namespace, name string) (map[GroupType][]string, bool)
+	// GetGroupsForExternalEntity returns the groups that select the given ExternalEntity.
+	GetGroupsForExternalEntity(namespace, name string) (map[GroupType][]string, bool)
+	// AddPod adds or updates a Pod to the index. If any existing groups are affected, eventHandlers will be called with
+	// the affected groups.
+	AddPod(pod *v1.Pod)
+	// DeletePod deletes a Pod from the index. If any existing groups are affected, eventHandlers will be called with
+	// the affected groups.
+	DeletePod(pod *v1.Pod)
+	// AddExternalEntity adds or updates an ExternalEntity to the index. If any existing groups are affected,
+	// eventHandlers will be called with the affected groups.
+	AddExternalEntity(ee *v1alpha2.ExternalEntity)
+	// DeleteExternalEntity deletes an ExternalEntity from the index. If any existing groups are affected, eventHandlers
+	// will be called with the affected groups.
+	DeleteExternalEntity(ee *v1alpha2.ExternalEntity)
+	// AddNamespace adds or updates a Namespace to the index. If any existing groups are affected, eventHandlers will be
+	// called with the affected groups.
+	AddNamespace(namespace *v1.Namespace)
+	// DeleteNamespace deletes a Namespace to the index. If any existing groups are affected, eventHandlers will be
+	// called with the affected groups.
+	DeleteNamespace(namespace *v1.Namespace)
+	// Run starts the index.
+	Run(stopCh <-chan struct{})
+	// HasSynced returns true if the interface has been initialized with the full lists of Pods, Namespaces, and
+	// ExternalEntities.
+	HasSynced() bool
+}
+
+type eventHandler func(group string)
+
+// entityType is an internal type used to differentiate Pod from ExternalEntity.
+type entityType int
+
+const (
+	podEntityType entityType = iota
+	externalEntityType
+)
+
+// GroupType is a public type used to differentiate Groups.
+type GroupType string
```

Comment:
nit for readability: this type definition would make more sense *before* defining the interface

Reply:
done